---
title: "82H: Stem Cell RNA-Seq Analysis Log"
author: "Emily Fradley"
date: "09/11/2020"
output:
  bookdown::pdf_document2: default
  bookdown::word_document2: default
  bookdown::html_document2: default
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      warning = FALSE,
                      message = FALSE,
                      fig.retina = 3)

##Create directories for work

dir.create("raw_data")
dir.create("processed_data")
dir.create("figures")
```

```{r packages, include=FALSE}
#Load packages
library(tidyverse)
library(bookdown)
library(citr)
library(GGally)
library(pheatmap)
library(scran)
library(DESeq2)
```

# Set Up, Tidy, and Summerise

We will be using R package `scran` [@scran] to analyse differences in expression data between LT-HSCs, HSPCs and Prog cells.

First we need to import the data, tidy the format and check it's structure to determine whether it is appropriate for the analysis we are going to do and to be able to interpret the results later. The data is in the form of 3 different data sets, one for each cell type.

```{r import}
#Import as matrix for findMarker to work, make sure to add row names
lthsc <- as.matrix(read.csv("raw_data/secretome_lthsc.csv", row.names=1))
hspc <- as.matrix(read.csv("raw_data/secretome_hspc.csv", row.names = 1))
prog <- as.matrix(read.csv("raw_data/secretome_prog.csv", row.names = 1))

```

## Data Structure

First we will check to see what the dimensions of the imported data looks like. Visually looking at the data shows us that each **gene is a row** and **each cell is a column**. We can find out the exact number of each using the code below:\
<!-- Summarizing some of the data to make ourselves more familiar -->

```{r data structure, echo=TRUE, results='hide'}
#Use ncol and nrow to extract the number of cells and genes respectively
#This can be double checked against the numbers reported by the str function

#LTHSCs number of cells and genes
ncol(lthsc)
nrow(lthsc)

#HSPCs number of cells and genes
ncol(hspc)
nrow(hspc)

#Prog number of cells and genes
ncol(prog)
nrow(prog)
```

So all the data sets have **`r nrow(hspc)` genes** or rows and the **number of cells** (i.e. columns) are respectively:\
- **LTHSCs: `r ncol(lthsc)`**\
- **HSPCs: `r ncol(hspc)`**\
- **Prog Cells: `r ncol(prog)`**

Next we can look to see if any of the genes or cells from the different data sets overlap using the `intersect` function.

```{r data overlap, include=FALSE}
overlap_row <- intersect(rownames(hspc), rownames(lthsc)) %>% length()
intersect(rownames(hspc), rownames(prog)) %>% length()
#423 overlapping genes, i.e. all genes are the same for all the data sets

overlap_col <- intersect(colnames(hspc), colnames(lthsc))
intersect(colnames(hspc), colnames(prog))
intersect(colnames(lthsc), colnames(prog))
#all returned character(0) therefore there is no overlap in the cells between the data sets

```

By comparing the rows and columns of each data set, we can see that the number of overlapping rows is `r overlap_row` for all 3 data sets. This is the same as the total number of rows of all the data sets so we have confirmed that all the data is showing **expression data for the same genes**.

For the overlap of the columns (i.e. the individual cells) we find that we have `r overlap_col` overlap between the data sets. This means **all the cells in each set are different from each other** and **different from the cells in the other data sets**.

Now we know that all the data sets contain the same `r overlap_row` genes but all have expression data for unique cells. This is what we expected.

------------------------------------------------------------------------

### Note

The original collection of the expression data contained over 2000 genes, our data has been cut down to include just over 400 genes. This has likely excluded any genes that showed no expression across the board of the different HSCs as any genes that weren't expressed would have been pointless to include in the analysis as we are looking to differentiate the cells based on differing expression patterns. There could be other reasons for the removal of this data though \*\*\*

## Linear or Log Data

To find genes of interest we are going to use the `findMarkers` function but to do that the expression data needs to be in **normalised log format**. The data has already been normalised and corrected for biases so all we need to do is check whether the data is in linear or log format.\
We can do this by checking the the maximum value of the expression data. We would expect linear data to return a very large number for this (in the thousands). In contrast we would expect log data to return a relatively small maximum value.

```{r check log, include=FALSE}
max_hspc <- max(hspc, na.rm = T)
max_lthsc <- max(lthsc, na.rm = T)
max_prog <- max(prog, na.rm = T)
```

The highest expression figures for the different data sets are as follows:\
HSPCs: `r max_hspc`\
LTHSC: `r max_lthsc`\
Prog Cells: `r max_prog`

This shows us that our results are most certainly log transformed and we can continue with the analysis of the data without having to transform it further.

## Summary Statistics

Finally we want to summarize the data we have for each gene and cell in a way that is easy to absorb (i.e. a plot). To do this we first have to generate some summary data including the mean and the standard deviation for each cell.

```{r summary-stats, include=FALSE}
#HSPC summary
hspc_mean_cell <- apply(hspc, 2, mean)
hspc_mean_gene <- apply(hspc, 1, mean)

hspc_sd_cell <- apply(hspc, 2, sd)
hspc_sd_gene <- apply(hspc, 1, sd)
         ###Combine into data.frame##
hspc_sum_cell <- data.frame(Mean=hspc_mean_cell, SD=hspc_sd_cell)
hspc_sum_gene <- data.frame(Mean=hspc_mean_gene, SD=hspc_sd_gene)
        ###Tidy up the Row names into columns so they can be called         and moved about
hspc_sum_cell <- cbind(Cell = rownames(hspc_sum_cell), hspc_sum_cell)
rownames(hspc_sum_cell) <- NULL

hspc_sum_gene <- cbind(Gene = rownames(hspc_sum_gene), hspc_sum_gene)
rownames(hspc_sum_gene) <- NULL

################################################################
#LTHSC
lthsc_mean_cell <- apply(lthsc, 2, mean)
lthsc_mean_gene <- apply(lthsc, 1, mean)

lthsc_sd_cell <- apply(lthsc, 2, sd)
lthsc_sd_gene <- apply(lthsc, 1, sd)
         ###Combine into data.frame##
lthsc_sum_cell <- data.frame(Mean=lthsc_mean_cell, SD=lthsc_sd_cell)
lthsc_sum_gene <- data.frame(Mean=lthsc_mean_gene, SD=lthsc_sd_gene)
    ###Tidy up the Row names into columns so they can be called         and moved about
lthsc_sum_cell <- cbind(Cell = rownames(lthsc_sum_cell), lthsc_sum_cell)
rownames(lthsc_sum_cell) <- NULL

lthsc_sum_gene <- cbind(Gene = rownames(lthsc_sum_gene), lthsc_sum_gene)
rownames(lthsc_sum_gene) <- NULL

##################################################################
#Prog cells
prog_mean_cell <- apply(prog, 2, mean)
prog_mean_gene <- apply(prog, 1, mean)

prog_sd_cell <- apply(prog, 2, sd)
prog_sd_gene <- apply(prog, 1, sd)

        ###Combine into data.frame##
prog_sum_cell <- data.frame(Mean=prog_mean_cell, SD=prog_sd_cell)
prog_sum_gene <- data.frame(Mean=prog_mean_gene, SD=prog_sd_gene)
    ###Tidy up the Row names into columns so they can be called         and moved about
prog_sum_cell <- cbind(Cell = rownames(prog_sum_cell), prog_sum_cell)
rownames(prog_sum_cell) <- NULL

prog_sum_gene <- cbind(Gene = rownames(prog_sum_gene), prog_sum_gene)
rownames(prog_sum_gene) <- NULL
```

```{r tidy-data, include=FALSE}
hspc_tidy <- hspc %>% 
              as_tibble(rownames='Gene') %>% 
              gather('Cell', 'Count', -Gene) %>% 
              mutate(CellType="HSPC")

lthsc_tidy <- lthsc %>% 
              as_tibble(rownames='Gene') %>% 
              gather('Cell', 'Count', -Gene) %>% 
              mutate(CellType="LTHSC")

prog_tidy <- prog %>% 
              as_tibble(rownames='Gene') %>% 
              gather('Cell', 'Count', -Gene) %>% 
              mutate(CellType="Progenetor")
```


## Visual Summary  
**NEED TO COME BACK TO THE FIRST PLOTS**  
After that we want to use the summary statistics to produce plots that will summarize the data sets for us and give us an idea of the overall expression patterns we have. We will use 2 plots:  

A good way to view the all of the expression data at once is to use a heat map. It's easier to digest visually but still gives us a good idea of expression patterns and early differences in the cell types.  


```{r heatmap, echo=TRUE}


```

The second graph is a parallel coordinate graph that we will use to plot the expression datas against each other.   

```{r coordinate fig}

```


# Results  

## findMarkers  
To use `findMarkers` we first need to combine the different cell type datasets so we can compare them. Then we apply the function to return statistical analysis results that tell use if there is difference between each genes expression in the different cell types. The results will return for each gene the **p-value**, **q (FDR) value**, and the **logFC (fold change)**.  
Now that we have these results saved we need to filter them to find our genes of interest.  

```{r combine data, include = FALSE}
#Bind 2 of the data matrices together
hspc_lthsc <- cbind(hspc, lthsc)
hspc_prog <- cbind(hspc, prog)
lthsc_prog <- cbind(lthsc, prog)

#check theyve bound properly by checking the structure
str(hspc_lthsc)
str(hspc_prog)
str(lthsc_prog)

#create labels for the different cell types
h_l_labels <- c(rep_len("HSPC", 701), rep_len("LTHSC", 155))
h_p_labels <- c(rep_len("HSPC", 701), rep_len("Prog", 798))
l_p_labels <- c(rep_len("LTHSC", 155), rep_len("Prog", 798))

```


```{r findmarkers, include = FALSE}
results_hl <- findMarkers(hspc_lthsc, h_l_labels)
results_hp <- findMarkers(hspc_prog, h_p_labels)
results_lp <- findMarkers(lthsc_prog, l_p_labels)
```

## Filtering Results 
### Statistical Significance

To determine which genes returned **statistically significant** results we need to filter them. This would usually done by p-value but since we have such a large data set we need to adjust the p-values to account for false positive results. The `findMarkers` function has already calculated the FDR (False Discovery Rate) or q value for us so now all we have to do is filter the results so we **only keep the genes with a FDR<0.01**. Simply this means we accept that the filtered results will include genes that are, at most, 1% false positives.  

```{r FDR, include=FALSE}
##  1% FDR 
sig01_hl <- results_hl$LTHSC[results_hl$LTHSC$FDR<0.01,]
sig01_hp <- results_hp$HSPC[results_hp$HSPC$FDR<0.01,]
sig01_lp <- results_lp$LTHSC[results_lp$LTHSC$FDR<0.01,]

## Number of genes that are significant:
nrow(sig01_hl)
nrow(sig01_hp)
nrow(sig01_lp)

##  5% FDR
sig05_hl <- results_hl$LTHSC[results_hl$LTHSC$FDR<0.05,]
sig05_hp <- results_hp$HSPC[results_hp$HSPC$FDR<0.05,]
sig05_lp <- results_lp$LTHSC[results_lp$LTHSC$FDR<0.05,]

## Number of genes that are significant:
nrow(sig05_hl)
nrow(sig05_hp)
nrow(sig05_lp)

##  10% FDR
sig10_hl <- results_hl$LTHSC[results_hl$LTHSC$FDR<0.10,]
sig10_hp <- results_hp$HSPC[results_hp$HSPC$FDR<0.10,]
sig10_lp <- results_lp$LTHSC[results_lp$LTHSC$FDR<0.10,]

## Number of genes that are significant:
nrow(sig10_hl)
nrow(sig10_hp)
nrow(sig10_lp)
```

For context we filtered the results with FDR of 1%, 5% and 10% to see the difference in the number of statistically significant genes we are left with. The results can be seen in the table below:

```{r FDR table}
FDRtab <- matrix(c(nrow(sig01_hl), nrow(sig05_hl), nrow(sig10_hl),
                     nrow(sig01_hp), nrow(sig05_hp), nrow(sig10_hp),
                     nrow(sig01_lp), nrow(sig05_lp), nrow(sig10_lp)),
                   ncol=3,byrow=TRUE)
colnames(FDRtab) <- c("1%","5%","10%")
rownames(FDRtab) <- c("HSPC/LTHSC","HSPC/Prog","LTHSC/Prog") 
FDRtab <- as.table(FDRtab)

knitr::kable(FDRtab,
  caption = "The number of genes that have a statistically significant difference in expression between the cell types HSPC & LTHSC, LTHSC & Progenitor, and LTHSC & Progenitors respectively when the maximum FDR value is set to 1%, 5%, and 10%.")
```

### Relative Differences

Next we need to take the results that we know are statistically significant and filter them by the actual difference in expression between cell populations. If the difference is too small it wouldn't matter if it is statistically significant or not as it will be unmeasurable in a practical setting and won't help us differ the cell types. The aim is to find genes that will act as indicators/markers for the cell type by their expression.

To asses the difference in the expression we use the `logFC` part of our `findMarkers` results. This will tell us the difference between our log2 expression values for the genes. That is a difference of 2 would represent a 2 fold (2x) increase in expression. 


### Absolute Expression
To help us get a better understanding of the actual expression levels to ensure we again have high enough expression to be measurable in practice we need to look at the genes **absolute expression**. We can do this by using the mean expressions of each gene that we already calculated earlier.

```{r abs exp, include =FALSE}
        ##HSPC/LTHSC

#sort the values so they are all in the same gene order
results_hl$HSPC <- results_hl$HSPC[match(names(hspc_mean_gene), row.names(results_hl$HSPC)),]

lthsc_mean_gene <- lthsc_mean_gene[match(names(hspc_mean_gene), names(lthsc_mean_gene))]

#then add the means as a new column to the results
#add means for both cell types
results_hl$HSPC$h_mean <- hspc_mean_gene
results_hl$HSPC$l_mean <- lthsc_mean_gene

      ##HSPC/Prog

#sort the values so they are all in the same gene order
results_hp$HSPC <- results_hp$HSPC[match(names(hspc_mean_gene), row.names(results_hp$HSPC)),]

prog_mean_gene <- prog_mean_gene[match(names(hspc_mean_gene), names(prog_mean_gene))]

#then add the means as a new column to the results
#add means for both cell types
results_hp$HSPC$h_mean <- hspc_mean_gene
results_hp$HSPC$p_mean <- prog_mean_gene

      ##LTHSC/Prog

#sort the values so they are all in the same gene order
results_lp$LTHSC <- results_lp$LTHSC[match(names(lthsc_mean_gene), row.names(results_lp$LTHSC)),]

prog_mean_gene <- prog_mean_gene[match(names(lthsc_mean_gene), names(prog_mean_gene))]

#then add the means as a new column to the results
#add means for both cell types
results_lp$LTHSC$l_mean <- lthsc_mean_gene
results_lp$LTHSC$p_mean <- prog_mean_gene

```

## Visual Results

To view both the statistical significance (p-value) and the the biological significance (log2fold change) at the same time we can use a volcano plot.

```{r volcano plot}
#alter results to dataframe
hl_results <- as.data.frame(results_hl$HSPC)
lp_results <- as.data.frame(results_lp$LTHSC)
hl_results$diffexpressed <- "NS"
hl_results$diffexpressed[hl_results$logFC.LTHSC > 1.5 & 
    hl_results$p.value < 0.05] <- "Over"
hl_results$diffexpressed[hl_results$logFC.LTHSC < -1.5 & hl_results$p.value < 0.05] <- "Under"


#plot volcanos
ggplot(data=hl_results, aes(x=logFC.LTHSC, y=-log10(p.value), col=diffexpressed)) + geom_point() + theme_minimal() +
  geom_vline(xintercept=c(-1.5, 1.5), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red") +
  scale_color_manual(values=c("grey", "pink", "lightblue"))

ggplot(data=lp_results, aes(x=logFC.Prog, y=-log10(p.value))) + geom_point() + theme_minimal()

```

We can also visualise results using an MA plot.

```{r MA plot, include = FALSE}
#create MA compliant dataframe
#hl_ma_results <- hl_results 
#hl_results[ -c(1,3:4) ]

```

## Differentially Expressed (DE) Genes

For this project we will be investigating **LT-HSC** which means we want to identify genes that are different between the LTHSCs and the other 2 cell types. 

```{r candidate genes, include=FALSE}
#candidate genes from H/L results
can1 <- filter(hl_results, p.value<0.05, FDR<0.01, summary.logFC<(-1)|summary.logFC>1)

#candidate genes from L/P results
can2 <- filter(lp_results, p.value<0.05, FDR<0.01, summary.logFC<(-1)|summary.logFC>1)

#candidate genes that are found in both groups
can_genes_names <- intersect(row.names(can1), row.names(can2))
can_genes_names
cat(can_genes_names, sep='\n')

can_genes <- can1[row.names(can1) %in% can_genes_names, ]
```

## Biological Investigation

Now we can use our list of differentially expressed (DE) genes and look at their biological function and orthologs to see if we can find any correlation between the statistically interesting genes and the biologically interesting ones. First we want to run a GO enrichment analysis. Since the genes of our original dataset were already filtered to only include secretome genes then we cannot assume that all the information given by the g:profiler software is inherent to our cell type. We must separate out that information from the information on enriched genes.


# References